crates/turbopack-ecmascript/src/resolve/mod.rs

```diff
@@ -41,23 +48,50 @@ pub async fn apply_cjs_specific_options(options: ResolveOptionsVc) -> Result<Res
 
 #[turbo_tasks::function]
 pub async fn esm_resolve(origin: ResolveOriginVc, request: RequestVc) -> Result<ResolveResultVc> {
-    let options = apply_esm_specific_options(origin.resolve_options());
-    specific_resolve(origin, request, options, "esm request").await
+    // TODO pass EcmaScriptModulesReferenceSubType
+    let ty = Value::new(ReferenceType::EcmaScriptModules(
+        EcmaScriptModulesReferenceSubType::Undefined,
+    ));
+    let options = apply_esm_specific_options(origin.resolve_options(ty.clone()));
+    specific_resolve(origin, request, options, ty).await
 }
 
 #[turbo_tasks::function]
 pub async fn cjs_resolve(origin: ResolveOriginVc, request: RequestVc) -> Result<ResolveResultVc> {
-    let options = apply_cjs_specific_options(origin.resolve_options());
-    specific_resolve(origin, request, options, "commonjs request").await
+    // TODO pass CommonJsReferenceSubType
+    let ty = Value::new(ReferenceType::CommonJs(CommonJsReferenceSubType::Undefined));
+    let options = apply_cjs_specific_options(origin.resolve_options(ty.clone()));
+    specific_resolve(origin, request, options, ty).await
+}
+
+#[turbo_tasks::function]
+pub async fn url_resolve(
+    origin: ResolveOriginVc,
+    request: RequestVc,
+    ty: Value<UrlReferenceSubType>,
+) -> Result<ResolveResultVc> {
+    let ty = Value::new(ReferenceType::Url(ty.into_value()));
+    let resolve_options = origin.resolve_options(ty.clone());
+    let rel_request = request.as_relative();
+    let rel_result = resolve(origin.origin_path().parent(), rel_request, resolve_options);
+    let result = if *rel_result.is_unresolveable().await? && rel_request.resolve().await? != request
+    {
+        resolve(origin.origin_path().parent(), request, resolve_options)
+            .add_references(rel_result.await?.get_references().clone())
+    } else {
+        rel_result
+    };
```

Comment:
Can you leave a comment explaining this? Why do we need to resolve a second time?